### ISSUE TITLE
!!! Data Curruption after adding records with PowerDNS-Admin !!!

### ISSUE THREAD

**Issue by schlitzered** [ngoduykhanh/PowerDNS-Admin#183]
Hey,

there is a severe bug that can corrupt your whole dataset of any zone.

how to reproduce:

1. Create a new zone example.com
2. open two tabs in this zone
3. in the first tab, create a new zone record, save & submit, verify that you can see the record, do not reload the second tab! 
4. in the second tab, that will not show this newly created record, create a second record, save & submit
5. reload both tabs: you will notice that the first entry is gone!

Analysis:
The problem is that all records, that where available while rendering the page, are send back, when you hit the submit button. The python application will then basically wipe the whole PowerDNS zone, and recreate all records that have been send via the webUI.

so if multiple users are editing the zone, they will overwrite each others changes!

also if a user is editing the zone directly via the PowerDNS REST API, or using a command line tool to edit the zone, PowerDNS-Admin will happily overwrite changes, if you do not have the most recent version of the zone rendered in your WebUI.

my solution to the problem would be to only send a delta via the WebUI, instead of all records originally rendered by the WebUI.

Kind Regards.

**Comment by trenb** [comment]
I've had this happen as well.

**Comment by phoenixlzx** [comment]
I think this is intended to be, you should never edit one zone in the same time.

According to PowerDNS API, you may sent only updated rrsets via `PATCH` method. However, this is not a delta because you don't know existing rrsets in the zone, and your records may still be overridden or corrupted.

IMO maybe simply set an 'edit lock' to the zone while entering edit mode, and removes the lock after submit or a timeout - just like the way that Wiki systems do.

**Comment by schlitzered** [comment]
no, this is IMHO wrong, PowerDNS Admin should only send updated/deleted records. for example, what if you also change records via dhcp ddns updates, what would be the reason to overwrite/delete these?

or imagine you have tens of thousands of records, sending them all the time is simply a stupid idea & very expensive.

i would call this unexpected behavior.

**Comment by phoenixlzx** [comment]
You can send a delta when you already get all the records, but this is not possible to make a delta or keep data integrity when the zone is edited simultaneously.

**Comment by schlitzered** [comment]
no, PowerDNS-Admin (the Javascript stuff) will not send a delta, it will always send the the full data set (since the page has been fetched), if you do not believe me, check it using firebug, or chrome dev tools.

after this PowerDNS-Admin, the python part, will:

- fetch a fresh copy of the zone records
- compare what has been send by Javascript with what is in PowerDNS
- if PowerDNS contains records that are not within what has been send by javascript, it will put these records in a list of records that should be removed, the list is called "postdata_for_delete"
-if PowerDNS contains records that are different to the ones that have been send by javascript, it will put these records in a list of records that should be updated, the list is called "postdata_for_new"
- now it will send these two lists to powerdns, 

this will remove records that have not been explicitly marked for deletion and this will update records that have not been explicitly marked for updated.

check app/models.py, line 900 to 1059 for the details


but maybe it is not clear what i mean with delta: with delta i mean, PowerDNS-Admin should only send records that have been modified/deleted by PowerDNS-Admin, records that have not been touched within the UI should not be set.

**Comment by jallakim** [comment]
Duplicate of #141 (different description/example, but the same underlying issue).

Only sending delta is not enough, since the API of pdns requires you to send *all* records of the same name and type in the same rrset.

If you have the following records;

```
foo.bar IN A 1.1.1.1
foo.bar IN MX 2.2.2.2
```

User #1 logs in and enters a zone, followed by idle. Meanwhile, user #2 enters the zone, and adds ```foo.bar IN A 3.3.3.3```, and saves. The updated zone would now have two A-records for foo.bar (1.1.1.1 + 2.2.2.2). User #1 comes back, and adds ```foo.bar IN A 4.4.4.4``` followed by a save. The updated zone would now have two A-records for foo.bar (1.1.1.1 + 4.4.4.4).

There is no way you could avoid this issue above with just sending deltas, as the delta (per the pdns API requirement) would be all of the A-records for foo.bar.

Hence, this issue should probably be solved in another way. The easy way would be to just disallow the update if the serial has changed (i.e. load the serial when loading the page, and upon submit, if the serial has changed, just discard all the changes and say "serial changed, please load the zone again"). Yes, it would "suck" to have all the changes discarded, but much better than what the current state is.

The hard way would be to implement some kind of diff-system within PowerDNS-Admin (i.e. before we send it to the pdns API), which would handle differences applied since you loaded the page, and make you able to resolve conflicts/changes since you loaded the page.

**Comment by Zuko** [comment]
https://github.com/zakk4223/PowerDNS-Admin/commit/544efc7fe36312a8b3c037e9457318dd9c4469c0 ?

**Comment by schlitzered** [comment]
checking the serial is not enough. adding/modifying /deleting records is not bound to increasing the serial.

if an cli, or api user makes changes, and is not increasing the serial, these changes will also be lost.

the optimal solution would be to only send the things that have been changed in the web-ui, and not every record. this will also increase performance when a zone has thousands of records.

**Comment by ngoduykhanh** [comment]
why adding/modifying /deleting does not increase serial? It is related to `SOA-EDIT-API
` setting?.

Sending only delta would increase the performance, however it is not supported by pdns API.

Even we can only send the things that have been changed in the web-ui, the problem still there i think. For example, you updating an existing record while other people removing it. What should be the final result?

**Comment by schlitzered** [comment]
at least with the CLI editing zones & increasing the serial are two, non related, tasks.

and yes, updating single entries (deltas) is supported by the pdns api, we are using this with "the foreman" for example. the foreman smart proxy only sends single entires, not the whole data set.

**Comment by jallakim** [comment]
I guess you have to send the entire FQDN when updating? I.e. if updating "blah.foo.bar", you have to take all the different record types into consideration (like, if it has A, AAAA and MX, you have to send all of them, even if you just updated the A record). Or am I mistaken?

**Comment by schlitzered** [comment]
please refer to the official documentation for this:

https://doc.powerdns.com/authoritative/http-api/zone.html#zone-endpoints

as far as i remember you simply only have to send the "rrsets" of the entries that have changes. and if i remember your code correctly you simply transmit an rrset for every record.
